### PR TITLE
api: handle redirect during websocket upgrade

### PR DESCRIPTION
**Pull request by lgfa29**
When attempting a WebSocket connection upgrade the client may receive a redirect request from the server, in which case the request should be reattempted using the new address present in the `Location` header.

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -986,28 +986,47 @@ func (c *Client) websocket(endpoint string, q *QueryOptions) (*websocket.Conn, *
...
-		var buf bytes.Buffer
+	if resp != nil {
+		switch resp.StatusCode {
+		case http.StatusSwitchingProtocols:
+			// Connection upgrade was successful.
+
+		case http.StatusPermanentRedirect, http.StatusTemporaryRedirect, http.StatusMovedPermanently:
+			locs := resp.Header["Location"]
+			if len(locs) == 0 {
+				return nil, nil, errors.New("redirect request missing Location header")
```
Comment (jrasell) [inline api/api.go:997]: Do we want to quote the `Location` entry to make it more obvious that's the header we are referring to?

File: api/api.go [hunk 2]
```diff
@@ -986,28 +986,47 @@ func (c *Client) websocket(endpoint string, q *QueryOptions) (*websocket.Conn, *
...
-		if resp.Header.Get("Content-Encoding") == "gzip" {
-			greader, err := gzip.NewReader(resp.Body)
+			loc := locs[0]
+			u, err := url.Parse(loc)
 			if err != nil {
-				return nil, nil, newUnexpectedResponseError(
-					fromStatusCode(resp.StatusCode),
-					withExpectedStatuses([]int{http.StatusSwitchingProtocols}),
-					withError(err))
+				return nil, nil, fmt.Errorf("invalid redirect location %s", loc)
```
Comment (jrasell) [inline api/api.go:1003]: Do we want to add the `err` string to the returned error?

File: api/api.go [hunk 3]
```diff
@@ -986,28 +986,47 @@ func (c *Client) websocket(endpoint string, q *QueryOptions) (*websocket.Conn, *
...
+
+			if resp.Header.Get("Content-Encoding") == "gzip" {
+				greader, err := gzip.NewReader(resp.Body)
+				if err != nil {
+					return nil, nil, newUnexpectedResponseError(
+						fromStatusCode(resp.StatusCode),
+						withExpectedStatuses([]int{http.StatusSwitchingProtocols}),
+						withError(err))
+				}
+				io.Copy(&buf, greader)
```
Comment (jrasell) [inline api/api.go:1018]:
```suggestion
				_, _ = io.Copy(&buf, greader)
```

File: api/api.go [hunk 4]
```diff
@@ -986,28 +986,47 @@ func (c *Client) websocket(endpoint string, q *QueryOptions) (*websocket.Conn, *
...
+			} else {
+				io.Copy(&buf, resp.Body)
```
Comment (jrasell) [inline api/api.go:1020]:
```suggestion
				_, _ = io.Copy(&buf, resp.Body)
```

File: api/api.go [hunk 5]
```diff
@@ -986,28 +986,47 @@ func (c *Client) websocket(endpoint string, q *QueryOptions) (*websocket.Conn, *
...
+			}
+			resp.Body.Close()
```
Comment (jrasell) [inline api/api.go:1022]:
```suggestion
			_ = resp.Body.Close()
```

**Review by jrasell — APPROVED** [review]
LGTM; just some minor comments and suggestions.

File: api/api.go [hunk 6]
```diff
@@ -986,28 +986,47 @@ func (c *Client) websocket(endpoint string, q *QueryOptions) (*websocket.Conn, *
...
-	if resp != nil && resp.StatusCode != http.StatusSwitchingProtocols {
-		var buf bytes.Buffer
+	if resp != nil {
+		switch resp.StatusCode {
+		case http.StatusSwitchingProtocols:
+			// Connection upgrade was successful.
+
+		case http.StatusPermanentRedirect, http.StatusTemporaryRedirect, http.StatusMovedPermanently:
+			locs := resp.Header["Location"]
+			if len(locs) == 0 {
```
Comment (shoenig) [inline api/api.go:996]: always use the `Header` methods which do MIME ... stuff

https://pkg.go.dev/net/http#Header.Get

Comment (lgfa29) [reply api/api.go:996]: Ah thanks!

**Pull request merged**